### PR TITLE
Add support for ICU message format

### PR DESCRIPTION
**Pull request by skress**
The ICU project (http://icu-project.org) provides a richer MessageFormat implementation than the JDK's default implementation.

This PR adds support for the ICU implementation. Default configuration keeps ICU format disabled, i.e. the JDK's default MessageFormat implementation is used. ICU message format can be enabled by setting play.i18n.useICUMessageFormat to true.

Please see MultiFormatMessagingPluginSpec for an example how to use the ICU message format.

### CODE REVIEW

**Comment by KarelCemus** [comment]
Hi @skress, 

first of all, thank you for your contribution. It is the great work.

However, I have several conceptual notes, which I don't like and want to discuss them with you.

> 1. `"com.ibm.icu" % "icu4j" % "62.1"` is always on the classpath even when the application does not use it

I propose to create another module, e.g., `play-i18n-icu`, which depends on both icu library and `play-i18n`, which implements the core. 

> 2. I very dislike that you make the icu as a key-value thing, it seems very violent and polutes both properties and yaml formats. 

Personally, I am not familiar with the ICU, though what is the most common file format for the localization? Your given example does not seem right. 

I'd propose to introduce another file format, e.g., `.icu` and have it there, although I believe there already exists some file format for ICU localization. That would also clean yaml and property files and would be explicitly obvious what the format is.

> 3. I don't like that `play.i18n.useICUMessageFormat` is `false` by default, while YAML and properties are true by default. This seems to be inconsistent as well as the configuration key itself. 

It would be better to be consistent with existing formats.

> 4. Implementation of the `translate` operation, which always checks the boolean and creates the `MessageFormat` seems very inefficient.

Since there already exists #13, which is similar to ICU I believe, then what about introduce a novel internal representation of the messages, which would replace the `Map`? It could be possibly a think wrapper, which would encapsulate one of possible implementations such as `MessageFormat` for ICU, similar thing for i18next, or just plain string in case of plain localization. It would be much more work but it would result in much cleaner and generic solution.

In conclusion and in order to deliver easy to use and generic library, I propose to update the `play-i18n` to provide a thin wrapper over internal representation of messages and define some injection points 
to inject additional factories for other file formats. Then, I propose to create a new module depending on `play-i18n`, which would use the file ICU format and implement the wrapper over `MessageFormat`.

What do you think?

@Enalmada Can you look at this PR and give us your feedback, please?

**Comment by skress** [comment]
Hi @KarelCemus 

thanks for your feedback. A couple of quick thoughts from my side:

> "com.ibm.icu" % "icu4j" % "62.1" is always on the classpath even when the application does not use it I propose to create another module, e.g., play-i18n-icu, which depends on both icu library and play-i18n, which implements the core.

You're right. Another possibility might be to declare icu4j as provided dependency. That might make it easier for users of this module to update icu4j independently of this module. 

> I very dislike that you make the icu as a key-value thing, it seems very violent and polutes both properties and yaml formats.
> Personally, I am not familiar with the ICU, though what is the most common file format for the localization? Your given example does not seem right.
> I'd propose to introduce another file format, e.g., .icu and have it there, although I believe there already exists some file format for ICU localization. That would also clean yaml and property files and would be explicitly obvious what the format is.

Well, it's not a key value thing. The key in the properties file is just "p.icu" for "this is a key-value pair in a properties file and it's value uses advanced features of the ICU format". The key could also just be "x.y.z". I only chose "p.icu" and "y.icu" to have a key-value pair in a properties file and in a YAML file for testing purposes.

> I don't like that play.i18n.useICUMessageFormat is false by default, while YAML and properties are true by default. This seems to be inconsistent as well as the configuration key itself.
> It would be better to be consistent with existing formats.

Do you have a concrete suggestion? I could name the property play.i18n.messageformat and values could be "default" or "icu".

>Implementation of the translate operation, which always checks the boolean and creates the MessageFormat seems very inefficient.
Neither ICU's MessageFormat is thread-safe nor the JDK's version. Play Framework's implementation of the translate method also always creates a new MessageFormat.

With regards to the boolean check: this also bothers me. Perhaps it would be better to have an additional class like MultiFormat**Icu**MessagesApi which does the same as MultiFormatMessagesApi but overrides the translate method.

To reduce code duplication this might need some more changes in the current implementation due to how the messages parameter is currently initialized.

>Since there already exists #13, which is similar to ICU I believe, then what about introduce a novel internal representation of the messages, which would replace the Map? It could be possibly a think wrapper, which would encapsulate one of possible implementations such as MessageFormat for ICU, similar thing for i18next, or just plain string in case of plain localization. It would be much more work but it would result in much cleaner and generic solution.

ICU MessageFormat can be used just like the JDK's MessageFormat, i.e. you provide numbered arguments in the form of a variable number of optional arguments after the property key. In addition to that ICU also supports named arguments which are provided as a `Map`.

> In conclusion and in order to deliver easy to use and generic library, I propose to update the play-i18n to provide a thin wrapper over internal representation of messages and define some injection points to inject additional factories for other file formats. Then, I propose to create a new module depending on play-i18n, which would use the file ICU format and implement the wrapper over MessageFormat.

In fact, the only thing that would be needed to support ICU would be to swap the implementation of `MessagesApi#translate` and that is totally orthogonal to the play-i18n module. Unfortunately that does not seem to be possible.

So I would suggest the above mentioned changes to keep things as simple as possible. If someone else then wants to provide another `MessageFormat` implementation (e.g. i18next) then that should be possible in the same way.

**Comment by KarelCemus** [comment]
> icu4j as provided dependency. 

That is one option but then it would need two maintain dependencies and their versions to make it work, which seems a bit overhead.

> That might make it easier for users of this module to update icu4j independently of this module.

Since it would be a transitive dependency within play-i18n, you can easily override it from the application. I wouldn't take this as an argument.

> Well, it's not a key value thing. The key in the properties file is just "p.icu" for "this is a key-value pair in a properties file and it's value uses advanced features of the ICU format". The key could also just be "x.y.z". I only chose "p.icu" and "y.icu" to have a key-value pair in a properties file and in a YAML file for testing purposes.

That is what I don't like. Why is there any key since it is not actually used and you have it only as a mandatory root. Instead, I'd propose to have an alternative file format `.icu`, which 1) does not have this root, 2) obviously declare its type in the file extension.

> Do you have a concrete suggestion? I could name the property play.i18n.messageformat and values could be "default" or "icu".

See [this](https://github.com/KarelCemus/play-i18n#configuration). There is, e.g., `play.i18n.formats.properties`, thus I'd use `play.i18n.formats.icu` to be consistent.

> Neither ICU's MessageFormat is thread-safe nor the JDK's version. Play Framework's implementation of the translate method also always creates a new MessageFormat.

I didn't know that, it is quite sad. I'd expect come immutable precompiled structure. Would it be possible (and made a sense) to implement it within play-i18n?

> With regards to the boolean check: this also bothers me. Perhaps it would be better to have an additional class like MultiFormatIcuMessagesApi which does the same as MultiFormatMessagesApi but overrides the translate method.

Yes, that's what I meant, besides other. However, I think a bit wider. I think there might exist multiple formats within a single project either due to legacy issues, migration, or just for simplicity. Therefore, it is necessary to support both ICU (or other) as well as plain strings. I think that having a different API implementation might be too broad and it might not be able to cover this situation. 

Play maintains localization pairs as a tuple key-value (String-String). I propose to replace this tuple by String-Wrapper, where the wrapper can be anything we need. For example, only a simple wrapper over a string in case of plain localization, but also a more complex structure internally using `MessageFormat`. The implementation of the wrapper could vary per a key, localization file respectively. This generic approach can be further extended to implement also additional formats such as i18next and possibly eventually also introduce extended `MessageApi` supporting plurals etc.

> only thing that would be needed to support ICU would be to swap the implementation of MessagesApi#translate and that is totally orthogonal to the play-i18n module. 

Is it really orthogonal? I am not much familiar with ICU but it does not seem so. I percieve it just as aditional input format just like YAML and properties. It is true that ICU is much more powerful but it is still "only additional format".

Am I missing something? What do you think? 

@Enalmada: You seem to be familiar with both i18next and ICU. What do you think? Any thoughts?

**Comment by skress** [comment]
ICU MessageFormat is only concerned with formatting a message after you already retrieved the message template. It does not matter if you got that template from a flat properties file, a hierarchical YAML file or a database.

I was searching for a way to use ICU MessageFormat with Play and I did not find anything. What I found was your extension and I thought it could be an easy change to add that functionality. Given our ongoing discussions it does not look to be such an easy change anymore ;-)

It might be interesting to follow your ideas about a much more generalised approach. Unfortunately I don't have enough time to go down that route. I've created a gist how to change the `MessagesApi#translate` implementation to use ICU's MessageFormat instead of the JDK's implementation. That's probably the easiest approach for anyone who is only interested in a different message pattern format and not in a different file structure for the organization of key-value pairs.

I will close this PR now. Thanks a lot for your time and for your plugin as it was a good inspiration for what I needed to do.

**Comment by KarelCemus** [comment]
I went again through your example (it is really complex, which confused me) and I believe I misunderstood ICU at first (I said I am not familiar with it). Now I see it is not a file format as I thought. However, I still think the solution should be a bit more complex than you proposed at first, possibly something you suggested in [this](https://github.com/KarelCemus/play-i18n/pull/15#issuecomment-419497671) comment.

- a separate module play-i18n-icu with proper dependencies as I suggested
- another implementation of MessagesAPI as you suggested

I think this is not that much implementation overhead as I proposed (as it does not make much sense since it is not a file format), but still, it needs slightly more work to keep the library maintainable and extensible in future.

However, I don't like your intended use with the `Map`:
``` icuMessages("p.icu", Map("gender" -> "male", "num_guests" -> 3, "host" -> "TheHost", "guest" -> "TheGuest", "d" -> date))(en)```

This is not much typesafe nor convenient and thus the Scala way, I think. What about introduction of `ICUMessagesApi` with apply method accepting varargs tuples instead of varargs, i.e., `(String,Any)*` instead of `Any*`? It is sligtly more typesafe since it requires tuples and definitely it is more convenient.
